conversion/resource_generation.go

```diff
@@ -0,0 +1,413 @@
+				ConnectionProfile: ConnectionProfile{
+					ProjectId:    projectId,
+					DatashardId:  profile.DataShardId,
+					Id:           profile.SrcConnectionProfile.Name,
+					IsSource:     true,
+					Host:         profile.SrcConnectionProfile.Host,
+					Port:         profile.SrcConnectionProfile.Port,
+					Password:     profile.SrcConnectionProfile.Password,
+					User:         profile.SrcConnectionProfile.User,
+					Region:       profile.SrcConnectionProfile.Location,
+					ValidateOnly: validateOnly},
+				ctx: ctx,
+			})
+		}
+		if !dstProfileExists && !validateOnly {
+			id := profile.DstConnectionProfile.Name
+			if id == "" {
+				id = "CN-" + uuid.New().String()
+				profile.DstConnectionProfile.Name = id
+			}
+			if profile.DstConnectionProfile.Location == "" {
+				profile.DstConnectionProfile.Location = region
+			}
+			dstProfilesToCreate = append(dstProfilesToCreate, &ConnectionProfileReq{
+				ConnectionProfile: ConnectionProfile{
+					ProjectId:    projectId,
+					DatashardId:  profile.DataShardId,
+					Id:           id,
+					IsSource:     false,
+					Region:       profile.DstConnectionProfile.Location,
+					BucketName:   "GCS-" + uuid.New().String(),
+					ValidateOnly: false},
+				ctx: ctx,
+			})
+		}
+	}
+	return sourceProfilesToCreate, dstProfilesToCreate, err
+}
+
+// 1. For each datashard, check if source and destination connection profile exists or not
+// 2. If source connection profile doesn't exists create it or validate if creation is possible.
+// 3. If validation is false and destination connection profile doesn't exists create a corresponding gcs bucket and then a destination connection profile
+func (r ResourceGenerationStruct) CreateResourcesForShardedMigration(ctx context.Context, projectId string, instanceName string, validateOnly bool, region string, sourceProfile profiles.SourceProfile, dsClient *datastream.Client) error {
+	var sourceProfilesToCreate []*ConnectionProfileReq
+	var dstProfilesToCreate []*ConnectionProfileReq
+
+	// Fetches list with resources which do not exist and need to be created
+	s := ResourceGenerationStruct{}
+	sourceProfilesToCreate, dstProfilesToCreate, err := s.getResourcesForCreation(ctx, projectId, sourceProfile, region, validateOnly, dsClient)
+	if err != nil {
+		return fmt.Errorf("resource generation failed %s", err)
+	}
+
+	// If validating resource creation, validate for all connection profiles. If creating, return error for the first resource creation that fails.
+	fastExit := false
+	if !validateOnly {
+		fastExit = true
+	}
+
+	var errorsList []error = []error{}
+
+	// Create or validate source connection profiles in parallel threads
+	resSourceProfiles, resCreationErr := common.RunParallelTasks(sourceProfilesToCreate, 20, r.prepareMinimalDowntimeResources, fastExit)
+	// If creation failed, perform cleanup of resources
+	if resCreationErr != nil && !validateOnly {
+		err = r.connectionProfileCleanUp(ctx, resourcesForCleanup)
```

Comment:
1. Please check with @VardhanThigle if this needs to be rate limited. We have this cause issues in the past.
2. Why isn't `connectionProfileCleanUp` also parallelised?